python_modules/dagster-graphql/dagster_graphql/implementation/fetch_asset_checks.py

```diff
@@ -102,3 +103,92 @@ def fetch_executions(
 
     return res
+
+
+def _execution_targets_latest_materialization(
+    instance: DagsterInstance,
+    external_asset_check: ExternalAssetCheck,
+    execution: AssetCheckExecutionRecord,
+    resolved_status: AssetCheckExecutionResolvedStatus,
+) -> bool:
+    # always show in progress checks
+    if resolved_status == AssetCheckExecutionResolvedStatus.IN_PROGRESS:
+        return True
+
+    records = instance.get_asset_records([external_asset_check.asset_key])
+    latest_materialization = records[0].asset_entry.last_materialization_record if records else None
+
+    if not latest_materialization:
+        # asset hasn't been materialized yet, so no reason to hide the check
+        return True
+
+    if resolved_status in [
+        AssetCheckExecutionResolvedStatus.SUCCEEDED,
+        AssetCheckExecutionResolvedStatus.FAILED,
+    ]:
+        evaluation = cast(
+            AssetCheckEvaluation,
+            check.not_none(
+                check.not_none(execution.evaluation_event).dagster_event
+            ).event_specific_data,
+        )
+        if not evaluation.target_materialization_data:
+            # check ran before the materialization was created
+            return False
+
+        # if the check matches the latest materialization, then show it
+        return (
+            evaluation.target_materialization_data.storage_id == latest_materialization.storage_id
+        )
+
+    # in this case the evaluation didn't complete, so we don't have target_materialization_data
+    elif resolved_status in [
+        AssetCheckExecutionResolvedStatus.EXECUTION_FAILED,
+        AssetCheckExecutionResolvedStatus.SKIPPED,
+    ]:
+        # if the check is executed in the same run as the materialization, then show it
+        latest_materialization_run_id = latest_materialization.event_log_entry.run_id
+        if latest_materialization_run_id == execution.run_id:
+            return True
+
+        # As a last ditch effort, check if the check's run was launched after the materialization's
+        latest_materialization_run_record = instance.get_run_record_by_id(
+            latest_materialization_run_id
+        )
+        execution_run_record = instance.get_run_record_by_id(execution.run_id)
+        return bool(
+            latest_materialization_run_record
+            and execution_run_record
+            and execution_run_record.create_timestamp
+            > latest_materialization_run_record.create_timestamp
+        )
+
+    else:
+        check.failed(f"Unexpected check status {resolved_status}")
+
+
+def fetch_execution_for_latest_materialization(
+    instance: DagsterInstance, external_asset_check: ExternalAssetCheck
+) -> Optional[GrapheneAssetCheckExecution]:
+    # we hide executions if they aren't for the latest asset materialization.
+    # currently we only consider the most recently launched check.
+
+    executions = instance.event_log_storage.get_asset_check_executions(
```

Comment:
This all looks great and +1 for pushing this logic to the backend so it's easier to change. Pretty nuanced stuff in here.

I wonder if we should also add an early return if the asset is partitioned since the "last" check could be for a backfill / not for the "latest" partition? I haven't looked at the latest mockups yet though, not sure what our intended UI is for the partitioned asset case.

Reply:
The current thinking is that checks act as if they have an all-to-one partition mapping (since checks can't be partitioned, yet). If any upstream partition is rematerialized, then the previous check result no longer applies

That said we haven't thought about this a bunch or gotten user feedback, so it could definitely change